src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1515 @@
+    return (NULL);
+  }
+
+  /* Allocate ARKodeLSRKStepMem structure, and initialize to zero */
+  step_mem = NULL;
+  step_mem = (ARKodeLSRKStepMem)malloc(sizeof(struct ARKodeLSRKStepMemRec));
+  if (step_mem == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_ARKMEM_FAIL);
+    ARKodeFree((void**)&ark_mem);
+    return (NULL);
+  }
+  memset(step_mem, 0, sizeof(struct ARKodeLSRKStepMemRec));
+
+  /* Attach step_mem structure and function pointers to ark_mem */
+  ark_mem->step_init              = lsrkStep_Init;
+  ark_mem->step_fullrhs           = lsrkStep_FullRHS;
+  ark_mem->step                   = lsrkStep_TakeStepRKC;
+  ark_mem->step_printallstats     = lsrkStep_PrintAllStats;
+  ark_mem->step_writeparameters   = lsrkStep_WriteParameters;
+  ark_mem->step_resize            = lsrkStep_Resize;
+  ark_mem->step_free              = lsrkStep_Free;
+  ark_mem->step_printmem          = lsrkStep_PrintMem;
+  ark_mem->step_setdefaults       = lsrkStep_SetDefaults;
+  ark_mem->step_getestlocalerrors = lsrkStep_GetEstLocalErrors;
+  ark_mem->step_mem               = (void*)step_mem;
+  ark_mem->step_supports_adaptive = SUNTRUE;
+
+  /* Set default values for optional inputs */
+  retval = lsrkStep_SetDefaults((void*)ark_mem);
+  if (retval != ARK_SUCCESS)
+  {
+    arkProcessError(ark_mem, retval, __LINE__, __func__, __FILE__,
+                    "Error setting default solver options");
+    ARKodeFree((void**)&ark_mem);
+    return (NULL);
+  }
+
+  /* Copy the input parameters into ARKODE state */
+  step_mem->fe = fe;
+  step_mem->fi = fi;
+
+  /* Update the ARKODE workspace requirements -- UPDATE */
+  ark_mem->liw += 0; /* fcn/data ptr, int, long int, sunindextype, sunbooleantype */
+  ark_mem->lrw += 0;
+
+  /* Initialize all the counters */
+  step_mem->nfe = 0;
+  step_mem->nfi = 0;
+
+  /* Initialize main ARKODE infrastructure */
+  retval = arkInit(ark_mem, t0, y0, FIRST_INIT);
+  if (retval != ARK_SUCCESS)
+  {
+    arkProcessError(ark_mem, retval, __LINE__, __func__, __FILE__,
+                    "Unable to initialize main ARKODE infrastructure");
+    ARKodeFree((void**)&ark_mem);
+    return (NULL);
+  }
+
+  /* Specify preferred interpolation type */
+  ark_mem->interp_type = ARK_INTERP_LAGRANGE;
+
+  return ((void*)ark_mem);
+}
+
+/*---------------------------------------------------------------
+  LSRKStepReInit:
+
+  This routine re-initializes the LSRKStep module to solve a new
+  problem of the same size as was previously solved. This routine
+  should also be called when the problem dynamics or desired solvers
+  have changed dramatically, so that the problem integration should
+  resume as if started from scratch.
+
+  Note all internal counters are set to 0 on re-initialization.
+  ---------------------------------------------------------------*/
+int LSRKStepReInit(void* arkode_mem, ARKRhsFn fe, ARKRhsFn fi, sunrealtype t0,
```

Comment:
Code duplication with create function. Extracting common functionality into a function may be useful